### DISCUSSION TITLE
aws - tagging - universal_augment backwards compatibility

### DISCUSSION THREAD

**Discussion by JohnHillegass** [cloud-custodian/cloud-custodian#7399] [General]
A few months ago, AWS added the capability to use ResourceARNList with the resourcegroupstagging api. This capability was introduced to custodian via #6414 and was introduced to botocore in version `1.19.58` https://github.com/boto/botocore/commit/1ca65a2baf5caef63bf33c13d3e2cadd43e82d8c. [Python lambda runtimes](https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html) use `botocore-1.19.31` which means we will need to add `packages: [ botocore]` to every cloudtrail based policy which is interacting with the resourcegroupstagging api to use the current version. 

While this is a good workaround, I think we could enable custodian to failsafe by checking for the 1.19.58 version in that method or catching the error and retrying using the previous approach. Thoughts?

Policy
```
policies:
  - name: sqs-net-new-test
    resource: sqs
    mode:
      role: arn:aws:iam::{account_id}:role/CloudCustodianRole
      type: cloudtrail
      events:
        - source: sqs.amazonaws.com
          event: CreateQueue
          ids: responseElements.queueUrl
    actions:
     - type: tag
       key: Found
       value: "true"
```

Error message
```
[ERROR]	2021-03-08T18:15:33.907Z	1e755386-590b-4957-9690-7df641bfd4f9	error during policy execution
Traceback (most recent call last):
  File "/var/task/c7n/handler.py", line 165, in dispatch_event
    p.push(event, context)
  File "/var/task/c7n/policy.py", line 1140, in push
    return mode.run(event, lambda_ctx)
  File "/var/task/c7n/policy.py", line 437, in run
    resources = self.resolve_resources(event)
  File "/var/task/c7n/policy.py", line 632, in resolve_resources
    return super().resolve_resources(event)
  File "/var/task/c7n/policy.py", line 419, in resolve_resources
    resources = self.policy.resource_manager.get_resources(resource_ids)
  File "/var/task/c7n/resources/sqs.py", line 95, in get_resources
    resources = super(SQS, self).get_resources(ids_normalized, cache)
  File "/var/task/c7n/query.py", line 560, in get_resources
    resources = self.augment(resources)
  File "/var/task/c7n/query.py", line 572, in augment
    return self.source.augment(resources)
  File "/var/task/c7n/resources/sqs.py", line 43, in augment
    return universal_augment(
  File "/var/task/c7n/tags.py", line 91, in universal_augment
    resource_tag_results = client.get_resources(
  File "/var/runtime/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/var/runtime/botocore/client.py", line 648, in _make_api_call
    request_dict = self._convert_to_request_dict(
  File "/var/runtime/botocore/client.py", line 696, in _convert_to_request_dict
    request_dict = self._serializer.serialize_to_request(
  File "/var/runtime/botocore/validate.py", line 297, in serialize_to_request
    raise ParamValidationError(report=report.generate_report())
botocore.exceptions.ParamValidationError: Parameter validation failed:
Unknown parameter in input: "ResourceARNList", must be one of: PaginationToken, TagFilters, ResourcesPerPage, TagsPerPage, ResourceTypeFilters, IncludeComplianceDetails, ExcludeCompliantResources
```

**Comment by JohnHillegass** [comment]
I am going back and forth on what alternatives to #6521 would be and have outlined some proposals below. @cloud-custodian/aws what do you think the best approach to solving this would be?

**Option 1**
handle the version difference via #6521

Pros:
- operators don't need to worry about it
- as python runtime is updated, this will self deprecate
- there is one obvious way to handle this specific bug

Cons:
- O(n) on tagging lambdas until the python runtime is updated

**Option 2**
force an override via package upload in the lambda

Pros:
- required packages can be toggled via a cli flag
- there is one obvious way to handle this class of bugs 

Cons:
- lambdas will become too large to see in the console and cannot be debugged as easily
- larger lambda size will increase the rate of storage exhaustion and require more frequent cleanup
- indeterminate amount of time that operators need to supply the package overrides

**Option 3**
force an override via package upload in a layer

Pros:
- required packages can be toggled via a cli flag
- reuse of a layer helps to keep lambda size down
- there is one obvious way to handle this class of bugs 

Cons:
- layer versioning and cleanup needs to be considered (increased management complexity)
- implications for layer version permission management (needs more investigation)
- indeterminate amount of time that operators need to supply the package overrides

**Comment by IngussNeilands** [comment]
hmmh, I am seeing these errors too, then I found this ticket yay
not sure how it is possible for custodian to get other than boto3==1.17.33
since it is hard a requirement in requirements.txt, but hey - we see these errors ...

**Comment by JohnHillegass** [comment]
Hey @IngussNeilands based off of my research its actually a botocore issue. @kapilt is thinking we can add a feature that deploys a lambda layer with the latest packages from c7n requirements so the lambdas are always current.

**Comment by IngussNeilands** [comment]
hmmh, what if bumping boto3 version to boto3>=1.19.58 helps ?
or opening it for us to overwrite like boto3>=1.17 ?
just thinking out loud

**Comment by IngussNeilands** [comment]
🤦 not boto3 but botocore ... right ...

**Comment by kapilt** [comment]
we should be frozen entirely wrt to pip install of custodian, on install.

the issue being refferenced here is auto packaging for lambda, at the moment you can do modules: [botocore, boto3] in your policy modes to get it, but the notion here is how can custodian do that automatically and also do that as a layer so users don't have to specify that.

we face this pain point everytime we use a newer sdk feature in lambda, till the lambda runtime version updates its boto sdk.
